src/core/collision.hpp

```diff
@@ -109,14 +101,31 @@ bool validate_collision_parameters();
 /** @brief Add the collision between the given particle ids to the collision
  *  queue
  */
-void queue_collision(int part1, int part2);
-
-/** @brief Check additional criteria for the glue_to_surface collision mode */
-inline bool glue_to_surface_criterion(Particle const &p1, Particle const &p2) {
-  return (((p1.p.type == collision_params.part_type_to_be_glued) &&
-           (p2.p.type == collision_params.part_type_to_attach_vs_to)) ||
-          ((p2.p.type == collision_params.part_type_to_be_glued) &&
-           (p1.p.type == collision_params.part_type_to_attach_vs_to)));
+void queue_collision(std::vector<int> particles);
+
+inline bool particle_type_criterion(Particle const &p1, Particle const &p2) {
+  int count_multiples = 1;
+  if (p1.p.type == p2.p.type)
+    count_multiples = 2;
+  if (std::count(collision_params.particle_type.begin(),
+                 collision_params.particle_type.end(),
+                 p1.p.type) == count_multiples)
+    return (count_multiples == 2 ||
+            std::count(collision_params.particle_type.begin(),
+                       collision_params.particle_type.end(), p2.p.type) > 0);
+  return 0;
+}
+
+inline bool collision_detection_criterion(Particle const &p1,
+                                          Particle const &p2) {
+  if (collision_params.rate > 0)
+    return d_random() < collision_params.rate * time_step;
```

Comment:
Function `d_random()` was removed in 0abe004dc28. You can either use a Philox counter (ex: [`BaseThermostat`](https://github.com/espressomd/espresso/blob/af75344bed302efeab0910234fad4cb4e956ae77/src/core/thermostat.hpp#L95-L121)), or a Mersenne Twister (ex: [`ParticleCache_test`](https://github.com/espressomd/espresso/blob/af75344bed302efeab0910234fad4cb4e956ae77/src/core/unit_tests/ParticleCache_test.cpp#L67-L76)), and pass their random integers to a function that generates a real number between 0 and 1 (ex: [`Utils::uniform()`](https://github.com/espressomd/espresso/blob/af75344bed302efeab0910234fad4cb4e956ae77/src/utils/include/utils/uniform.hpp#L36-L41)). In both cases, you'll need to store the RNG state somewhere.